File: cpp/Canvas.cpp
#include "Canvas.h"

Canvas::Canvas(int width, int height): width_{width}, height_{height}, pixels_(width * height) {
}

void Canvas::PutPixel(int x, int y, const Color& color) {
    x += (width() >> 1);
    y += (height() >> 1);
    if (x < 0 || x >= width() || y < 0 || y >= height()) {
        return;
    }
    pixels_[y * width_ + x] = color;
}

const Color& Canvas::GetPixel(int x, int y) const {
    return (*this)[y * width_ + x];
}

const Color& Canvas::operator[](size_t idx) const {
    return pixels_[idx];
}

int Canvas::width() const {
    return width_;
}

int Canvas::height() const {
    return height_;
}

File: rust/src/canvas.rs
use crate::color::Color;

pub struct CanvasPosition {
    pub x: i32,
    pub y: i32,
}

pub enum ImagePosition {
    Canvas { x: i32, y: i32 },
    Screen { x: u32, y: u32 },
}

pub trait Drawable {
    fn width(&self) -> u32;
    fn height(&self) -> u32;
    fn put_pixel(&mut self, position: ImagePosition, color: Color);
}
